matrix_sdk/src/room/common.rs

```diff
@@ -105,45 +107,163 @@ impl Common {
-    /// returns a `get_message_events::Response` that contains a chunk of
-    /// room and state events (`AnyRoomEvent` and `AnyStateEvent`).
+    /// Gets a slice of the timeline of this room
+    ///
+    /// Returns a slice of the timeline between `start` and `end`, no longer
+    /// then `limit`. If the number of events is fewer then `limit` it means
+    /// that in the given direction no more events exist.
+    /// If the timeline doesn't contain an event with the given `start` `None`
+    /// is returned.
     ///
     /// # Arguments
     ///
-    /// * `request` - The easiest way to create this request is using the
-    /// `get_message_events::Request` itself.
+    /// * `start` - An `EventId` that indicates the start of the slice.
+    ///
+    /// * `end` - An `EventId` that indicates the end of the slice.
+    ///
+    /// * `limit` - The maximum number of events that should be returned.
+    ///
+    /// * `direction` - The direction of the search and returned events.
     ///
     /// # Examples
     /// ```no_run
     /// # use std::convert::TryFrom;
     /// use matrix_sdk::Client;
-    /// # use matrix_sdk::ruma::room_id;
+    /// # use matrix_sdk::ruma::{event_id, room_id};
     /// # use matrix_sdk::ruma::api::client::r0::{
     /// #     filter::RoomEventFilter,
-    /// #     message::get_message_events::Request as MessagesRequest,
+    /// #     message::get_message_events::{Direction, Request as MessagesRequest},
     /// # };
     /// # use url::Url;
     ///
     /// # let homeserver = Url::parse("http://example.com").unwrap();
     /// let room_id = room_id!("!roomid:example.com");
-    /// let request = MessagesRequest::backward(&room_id, "t47429-4392820_219380_26003_2265");
     ///
     /// let mut client = Client::new(homeserver).unwrap();
     /// # let room = client
     /// #    .get_joined_room(&room_id)
     /// #    .unwrap();
     /// # use futures::executor::block_on;
     /// # block_on(async {
-    /// assert!(room.messages(request).await.is_ok());
+    /// assert!(room.messages(&event_id!("$xxxxxx:example.org"), None, 10, Direction::Backward).await.is_ok());
     /// # });
     /// ```
     pub async fn messages(
         &self,
-        request: impl Into<get_message_events::Request<'_>>,
-    ) -> Result<get_message_events::Response> {
-        let request = request.into();
-        self.client.send(request, None).await
+        start: &EventId,
+        end: Option<&EventId>,
+        limit: usize,
+        direction: Direction,
+    ) -> Result<Option<Vec<SyncRoomEvent>>> {
+        let room_id = self.inner.room_id();
+        let events = if let Some(mut stored) = self
+            .client
+            .store()
+            .get_timeline(room_id, Some(start), end, Some(limit), direction.clone())
+            .await?
+        {
+            // We found a gab or the end of the stored timeline.
+            if let Some(token) = stored.token {
```

Comment:
Making the request every time is what keeps the cache from ever going stale and it also is what prevents gaps right? It's too bad there isn't a more straightforward way to avoid the request every time.

Could the [/message](https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-rooms-roomid-messages) endpoint be used for a more exact finding of the missing events? It would avoid needing to double the limit and could use the given `direction`.

Reply:
That comment at 165 should also say "gap" instead of "gab".

Reply:
> Making the request every time is what keeps the cache from ever going stale and it also is what prevents gaps right? It's too bad there isn't a more straightforward way to avoid the request every time.

No, the request is made only when not enough events where found in the store.

Reply:
> it also is what prevents gaps right?
Not really, but if this hits a gap we fill it with the request. In the specific case that the gap was only less then `limit` we are requesting messages that are actually stored in the store but that's impossible to know before making the request since the size of the gap is only known once it doesn't exist any more, I mean once we filled it.

Reply:
@DevinR528 how would the cache get stale?

Is it possible that the same message requests, with the same two tokens, return different results?

Reply:
I don't think it is possible for /sync to do that because they send the event in the order they are received but, /messages or /get_context might??? The spec says `in chronological order` but I think that just means small -> big and ordered by topological sort in which case I think a new event could slip in between two old/seen events?

Spec:
> Events are ordered in this API according to the arrival time of the event on the homeserver. This can conflict with other APIs which order events based on their partial ordering in the event graph.

```
a                    b
A ---> B ---> C ---> D
New event comes in from a previously non responsive server
a                            b
A ---> B ---> B' ---> C ---> D
```

Reply:
hmm, so the new events (that comes in from a previously non responsive server) are included in a sync response, which will cause a different ordering for the event in the local timeline, but IMO as long as we include all events the specific ordering isn't that important since even different server can have different ordering of the timeline.

Reply:
Well the sync response would have a completely set and new token, to continue the example it would be sync response one = `A, B, C, D` as `tkn_1` to `tkn_2` then sync response 2 = `B'` using a new `tkn_3` to `tkn_4` the problem would be if someone used /messages or /get_context, so if we say our `Common::messages` is the same as calling /messages or eventually if we add a `Common::get_context` the cases when the call makes requests and the case where it uses the cache would differ. Having a `start/end` eventId would also be arbitrary (as in they aren't necessarily sequential).

> since even different server can have different ordering of the timeline.

While this may be basically true in practice sort of, the idea is that all servers do agree eventually (eventually consistent).

I do wonder if we just kind of go with what you have (because it is really good IMHO!!) if we could say "our timeline store is ordered the same way sync is ordered for the most part" and as long as we don't have duplicates and keep the chunks together.

Reply:
Sounds good to me. I also added a method to clear the timeline so that it will be fetched again.